packages/util/src/polymorphicForwardRef.ts

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (c) 2023-present, NDLA.
+ *
+ * This source code is licensed under the GPLv3 license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import {
+  ComponentProps,
+  ComponentPropsWithoutRef,
+  ElementType,
+  ForwardRefRenderFunction,
+  ValidationMap,
+  WeakValidationMap,
+  forwardRef,
+} from 'react';
+
+export type PropsOf<T extends ElementType> = ComponentPropsWithoutRef<T> & {
+  element?: ElementType;
+};
+
+export type RightJoinProps<SourceProps extends object = {}, OverrideProps extends object = {}> = Omit<
+  SourceProps,
+  keyof OverrideProps
+> &
+  OverrideProps;
+
+export type MergeWithAs<
+  ComponentProps extends object,
+  AsProps extends object,
+  AdditionalProps extends object = {},
+  AsComponent extends ElementType = ElementType,
+> = (RightJoinProps<ComponentProps, AdditionalProps> | RightJoinProps<AsProps, AdditionalProps>) & {
+  element?: AsComponent;
+};
+
+export type ComponentWithAs<Component extends ElementType, Props extends object = {}> = {
+  <AsComponent extends ElementType = Component>(
+    props: MergeWithAs<ComponentProps<Component>, ComponentProps<AsComponent>, Props, AsComponent>,
+  ): JSX.Element;
+  displayName?: string;
+  propTypes?: WeakValidationMap<any>;
+  contextTypes?: ValidationMap<any>;
+  defaultProps?: Partial<any>;
+  id?: string;
+};
+
+export type PolymorphicProps<T extends ElementType> = Omit<PropsOf<T>, 'ref'> & { element?: ElementType };
+
+/**
+ * A super complicated type that allows us to wrap polymorphic components in a forwardRef.
+ * Taken from [Chakra-ui](https://github.com/chakra-ui/chakra-ui/blob/main/packages/core/system/src/forward-ref.tsx)
```

Comment:
Bør vi ha med lisensen til chakra om det er tilfelle?
Ser de har MIT og den krever vel at du legger med lisens?
Oog bør denne url'en være en som peker på en commit sånn at den ikke brekker om de endrer på det i fremtiden?

Reply:
Når det kommer til lisensgivning har du mer peiling enn meg.

Akkurat denne funksjonen er såpass viktig for chakra at jeg sliter med å se for meg at den kommer til å forsvinne. Kunne eventuelt lagt inn en lenke til som lenker til den faktiske commit-hashen jeg brukte til å hente typene?

Reply:
> Når det kommer til lisensgivning har du mer peiling enn meg.

Om mulig så tror jeg det er ryddig å dra ut det vi har stjelt i en egen fil og hive med MIT lisensen til chakra. Da har vi i alle fall gjort vårt beste :smile: 

> Akkurat denne funksjonen er såpass viktig for chakra at jeg sliter med å se for meg at den kommer til å forsvinne. Kunne eventuelt lagt inn en lenke til som lenker til den faktiske commit-hashen jeg brukte til å hente typene?

Ja, det tenker jeg er lurt. 
Feks bruke denne. https://github.com/chakra-ui/chakra-ui/blob/86dd065351d4c5b6a612ae804dbf09cd7fe45799/packages/core/system/src/forward-ref.tsx

Taper ingenting på det, og det _kan_ jo skje at de har restrukturert noe kode e.l så den linken dør om 4 år når det kommer en kavring for å se på det her.